hls-plugin-api/src/Ide/Types.hs

```diff
@@ -792,13 +775,39 @@ type PluginNotificationMethodHandler a m = a -> VFS -> PluginId -> MessageParams
 
 -- | Make a handler for plugins with no extra data
 mkPluginHandler
-  :: PluginRequestMethod m
+  :: forall ideState m. PluginRequestMethod m
   => SClientMethod m
   -> PluginMethodHandler ideState m
   -> PluginHandlers ideState
-mkPluginHandler m f = PluginHandlers $ DMap.singleton (IdeMethod m) (PluginHandler f')
+mkPluginHandler m f = PluginHandlers $ DMap.singleton (IdeMethod m) (PluginHandler (f' m))
   where
-    f' pid ide params = pure <$> f ide pid params
+    f' :: SMethod m -> PluginId -> ideState -> MessageParams m -> LspT Config IO (NonEmpty (Either ResponseError (MessageResult m)))
+    -- We need to have separate functions for each method that supports resolve, so far we only support CodeActions
+    -- CodeLens, and Completion methods.
+    f' SMethod_TextDocumentCodeAction pid ide params@CodeActionParams{_textDocument=TextDocumentIdentifier {_uri}} =
+      pure . fmap (wrapCodeActions pid _uri) <$> f ide pid params
+    f' SMethod_TextDocumentCodeLens pid ide params@CodeLensParams{_textDocument=TextDocumentIdentifier {_uri}} =
+      pure . fmap (wrapCodeLenses pid _uri) <$> f ide pid params
+    f' SMethod_TextDocumentCompletion pid ide params@CompletionParams{_textDocument=TextDocumentIdentifier {_uri}} =
+      pure . fmap (wrapCompletions pid _uri) <$> f ide pid params
```

Comment:
Hmmm... couldn't we push this stuff into the helper functions we have for _making_ such handlers? Then, yes, you need to use those if you want to get resolve working, but if you don't care you can just write a bare handler and we won't insert extra stuff if you haven't asked for it?

Reply:
Right, so I am suggesting something just like that, I was even thinking more specific: `mkCodeLensHandlerWithResolve` or something. I don't think it's so bad for people to have to call a specific function to opt in to the magic?